design/handle-backup-of-volumes-by-resources-filters.md

```diff
@@ -0,0 +1,211 @@
+# Handle backup of volumes by resources filters
+ 
+## Abstract
+Currently, Velero doesn't have one flexible way to filter volumes.
+ 
+If users want to skip backup of volumes or only backup some volumes in different namespaces in batch, currently they need to use the opt-in and opt-out approach one by one, or use label-selector but if it has big different labels on each different related pods, which is cumbersome when they have lots of volumes to handle with. it would be convenient if Velero could provide one way to filter the backup of volumes just by `some specific volumes attributes`.
+ 
+Also, currently, it's not accurate enough if the users want to select a specific volume to do a backup or skip by without patching labels or annotations to the pods. It would be useful if users could accurately select target volume by `one specific resource selector`. Users could accurately select the volume to backup or skip in their own console when using velero for secondary development.
+ 
+## Background
+As of Today, Velero has lots of filters to handle (backup or skip backup) resources including resources filters like `IncludedNamespaces, ExcludedNamespaces`, label selectors like `LabelSelector, OrLabelSelectors`, annotation like `backup.velero.io/must-include-additional-items` etc. But it's not enough flexible to handle volumes, we need one generic way to filter volumes.
+ 
+## Goals
+- Introducing one flexible way to filter volumes.
+ 
+## Non Goals
+- We only handle volumes for backup and do not support restore.
+- Currently, only handle volumes, does not support other resources.
+- Only environment-unrelated and platform-independent general volumes attributes are supported, do not support volumes attributes related to a specific environment.
+ 
+## Use-cases/Scenarios
+### A. Skip backup volumes by some attributes
+Users want to skip PV with the requirements:
+- option to skip all PV data
+- option to skip specified PV type (RBD, NFS)
```

Comment:
indeed, the above description is made by PM, the precise description should be `option to skip specified PV volume sources`, which could be retrieved by `PersistentVolumeSource` of PV

All the PV attributes could be retrieved in `PersistentVolumeSpec` if the user could follow the Kubernetes PersistentVolume structure design.

Reply:
What about the CSI source? Behind the CSI driver, the storage type may also be NFS, RBD, etc.

Reply:
In my understanding, NFS can be used in Pod by three ways
- mount as Volume in Pod spec
-  mount as NFS type PV/PVC
- mount using CSI type PV/PVC where NFS CSI driver used

Is it not possible to detect volume source as NFS in all three cases? 

I have seen most time user wants to skip NFS volume in backup or want to skip such volumes from backup which does not support volume snapshot like NFS CSI driver does not support CSI volume snapshot.

In brief, user needs a way to say
- skip backup of all volumes & PV/PVC
- skip such volume & PV/PVC which does not support volumesnaphot AND backup all volumes which supports volumesnapshot

We would also need a way to specify:
- backup Volumes/PV using Kopia as well as volume snapshot
- backup  Volumes/PV using Kopia only (skip volume snapshot for volumes even it support volume snapshot)
- backup Volumes/PV using volume snapshot only (and skip such volumes which does not support volume snapshot)
- backup all Volumes/PV but use snapshot for volumes support snapshotting and backup rest volume using kopia

Reply:
> In brief, user needs a way to say

> skip backup of all volumes & PV/PVC
>skip such volume & PV/PVC which does not support volumesnaphot AND backup all volumes which supports volumesnapshot

we could skip by the volumes sources or storage class.

Reply:
@pradeepkchaturvedi 

I agree that we should let the user choose to use fs-based backup like kopia when the default approach is "volume snapshot" or "CSI snapshot". IMO in this design, we should make it possible to support an "if... then..." semantic so that the user can define if a PV meets certain conditions it can be backed up via certain ways. 

I don't think it's a high-priority use case to back up one object using different approaches, it makes the restore workflow more complicated and more confusing, and the user won't tell which way is better.

Reply:
> In my understanding, NFS can be used in Pod by three ways
> 
> * mount as Volume in Pod spec
> * mount as NFS type PV/PVC
> * mount using CSI type PV/PVC where NFS CSI driver used

@pradeepkchaturvedi @reasonerjt  
above are references to two different[ storage types](https://kubernetes.io/docs/concepts/storage/):
mount as Volume in Pod spec --> [pod volume](https://kubernetes.io/docs/concepts/storage/volumes/)
mount as NFS type PV/PVC --> PV
mount using CSI type PV/PVC where NFS CSI driver used --> PV

For our Velero to do the backup, we only needed to consider `pod volume` and `PV` to filter volumes.

How do we classify our resource filter configuration?
- classify by storage types: `pod volume` and `PV`
- classify by approach, which would be `Volume Snapshot`, `CSI snapshot`, `fs-based backup`, or even `data-mover`

obviously, classifying by storage types will be more clear. `pod volume` and `PV` has the same attribute volume source, which indicates the Volume type is `NFS` or `CSI` etc. which may cover most requirement. 


>I agree that we should let the user choose to use fs-based backup like kopia when the default approach is "volume snapshot" or "CSI snapshot". IMO in this design, we should make it possible to support an "if... then..." semantic so that the user can define if a PV meets certain conditions it can be backed up via certain ways.

Suppose we use filter configuration classified by storage types, If we don't follow the default approach specified by the command of `velero create`, instead let `PV meets certain conditions it can be backed up via certain ways`, which may be indeterminate. And it will bring complexity to the whole backup workflow. What about we only follow the default approach to handle PV meets certain conditions, if PV couldn't be handled by default approach just failed directly.  For example: if a user wants to backup a volume with a CSI snapshot, and PV has the CSI persistent volume source, but the CSI driver does not support snapshots for the underlying storage system does not support snapshots. maybe it's not easy for us to confirm the "if... then...".

Reply:
@qiuming-best 
IMO We should first support filtering "storage type" based on storage class, this may be sufficient already
https://github.com/vmware-tanzu/velero/pull/5773/files#r1090342124

>  And it will bring complexity to the whole backup workflow.
It will, but the user can choose not to use it.  IMO letting the user to choose backup via podvolume or snapshot is a valid requirement(we probably don't need to support it in v1.11, but we need to make sure the data structure has such flexibility).